### PR TITLE
[FEATURE] Add LAMB optimizer

### PR DESCRIPTION
**Pull request by kenjewu**
## Description ##
The LAMB optimizer: 
It has been proposed in [Reducing BERT Pre-Training Time from 3 Days to 76 Minutes](https://arxiv.org/abs/1904.00962).

A simple neural network has been used for verification, and the results show that the model can converge normally with large batchsize. Need to be verified further on the BERT.

@eric-haibin-lin  Verification may require your help because of some computing resource limitations.

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-733/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-733/1/index.html

**Comment by mli** [comment]
Job PR-733/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-733/2/index.html

**Comment by mli** [comment]
Job PR-733/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-733/3/index.html

**Comment by mli** [comment]
Job PR-733/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-733/4/index.html

File: src/gluonnlp/optimizer/lamb.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        mean, var = state
+        mean[:] = self.beta1 * mean + (1. - self.beta1) * grad
+        var[:] = self.beta2 * var + (1. - self.beta2) * square(grad)
+
+        # execution bias correction
+        mean_hat = mean / (1. - power(self.beta1, t))
+        var_hat = var / (1. - power(self.beta2, t))
+
+        r1 = weight.norm()
+        g = mean_hat / sqrt(var_hat + self.epsilon) + wd * weight
```
Comment (szhengac) [inline src/gluonnlp/optimizer/lamb.py]: put epsilon outside the `sqrt`

File: src/gluonnlp/optimizer/lamb.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        r2 = g.norm()
+
+        # calculate lamb_trust_ratio
+        r = 1. if r1 == 0. or r2 == 0. else minimum(
+            maximum(r1 / r2, self.lower_bound), self.upper_bound)
```
Comment (szhengac) [inline src/gluonnlp/optimizer/lamb.py]: the clip func only performs on r1  and g is normalized by r2.

File: src/gluonnlp/optimizer/lamb.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        if self.clip_gradient is not None:
+            grad = clip(grad, -self.clip_gradient, self.clip_gradient)
+
+        mean, var = state
+        mean[:] = self.beta1 * mean + (1. - self.beta1) * grad
+        var[:] = self.beta2 * var + (1. - self.beta2) * square(grad)
+
+        # execution bias correction
+        mean_hat = mean / (1. - power(self.beta1, t))
+        var_hat = var / (1. - power(self.beta2, t))
```
Comment (szhengac) [inline src/gluonnlp/optimizer/lamb.py]: it seems that bias correction is not performed in the algorithm. needs to double check.

File: src/gluonnlp/optimizer/lamb.py (hunk 1 above)
Comment (kenjewu) [reply src/gluonnlp/optimizer/lamb.py]: I compared the calculation formula in the original paper and found that epsilon is in sqrt.

Comment (szhengac) [reply src/gluonnlp/optimizer/lamb.py]: they change the algorithm in their newest arxiv version.

Comment (kenjewu) [reply src/gluonnlp/optimizer/lamb.py]: thanks for your reminder. I found that the paper I referenced was not synchronized with the latest one. I will update the relevant code later.

File: src/gluonnlp/optimizer/lamb.py (hunk 3 above)
Comment (xcgoner) [reply src/gluonnlp/optimizer/lamb.py]: What about we keep both versions of LAMB, and test which one is better by ourselves? There are significant differences between these 2, especially the existence of bias correction. Just add a flag so that the users can choose which one to use.

Comment (kenjewu) [reply src/gluonnlp/optimizer/lamb.py]: Two versions have been controlled using parameters,`use_latest`. I have tested two versions on a small model and there is no obvious difference. I think it should be tested on larger models and data.

**Comment by mli** [comment]
Job PR-733/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-733/5/index.html

File: src/gluonnlp/optimizer/lamb.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    Updates are applied by::
+    if use_latest:
+        grad = clip(grad * rescale_grad, clip_gradient)
+        m = beta1 * m + (1 - beta1) * grad
+        v = beta2 * v + (1 - beta2) * (grad**2)
+        r1 = minimum(maximum(w.norm(), self.lower_bound), self.upper_bound)
+        g = m / (sqrt(v_hat) + epsilon) + wd * w
+        r2 = g.norm()
+        r = if r1 == 0. or r2 == 0. else r1/r2
```
Comment (xcgoner) [inline src/gluonnlp/optimizer/lamb.py]: r = 1 if ...

File: src/gluonnlp/optimizer/lamb.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    else:
+        grad = clip(grad * rescale_grad, clip_gradient)
+        m = beta1 * m + (1 - beta1) * grad
+        v = beta2 * v + (1 - beta2) * (grad**2)
+        m_hat = m / (1 - power(beta1, t))
+        v_hat = m / (1 - power(beta2, t))
+        r1 = w.norm()
+        g = m_hat / (sqrt(v_hat + epsilon)) + wd * w
+        r2 = g.norm()
+        r = if r1 == 0. or r2 == 0. else r1/r2
```
Comment (xcgoner) [inline src/gluonnlp/optimizer/lamb.py]: r = 1 if ...

**Comment by mli** [comment]
Job PR-733/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-733/6/index.html

File: src/gluonnlp/optimizer/lamb.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    upper_bound : float, optional, default is 10.0
+        Upper limit of norm of weight
+    use_latest : bool, optional, default is True
+        Whether to use the latest version of LAMB. The new version of LAMB
+        has some differences from the old version,
+        such as the bias correction was removed in the new version
+    """
+
+    def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-6,
+                 lower_bound=1e-3, upper_bound=10.0, use_latest=True,
```
Comment (eric-haibin-lin) [inline src/gluonnlp/optimizer/lamb.py]: `use_latest` is not a good name. They might update their paper again. What about `bias_correction=False`?

File: src/gluonnlp/optimizer/lamb.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+from mxnet.optimizer import Optimizer, register
+from mxnet.ndarray import zeros, NDArray
+from mxnet.ndarray import square, power, sqrt, maximum, minimum, clip
+
+__all__ = ['LAMB']
+
+
+@register
+class LAMB(Optimizer):
+    """The LAMB optimizer proposed in
```
Comment (eric-haibin-lin) [inline src/gluonnlp/optimizer/lamb.py]: I just realized that gluonnlp.optimizer does not appear in the website. @kenjewu could you add that?

File: src/gluonnlp/optimizer/lamb.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.epsilon = epsilon
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        self.lazy_update = lazy_update
+        self.use_latest = use_latest
+
+    def create_state(self, index, weight):
+        stype = weight.stype if self.lazy_update else 'default'
```
Comment (eric-haibin-lin) [inline src/gluonnlp/optimizer/lamb.py]: `stype = weight.stype`

File: src/gluonnlp/optimizer/lamb.py (hunk 6 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/optimizer/lamb.py]: remove the `lazy_update ` argument

File: src/gluonnlp/optimizer/lamb.py (hunk 8 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/optimizer/lamb.py]: We probably need to update the implementation for mixed precision (with a copy of master weight)@xcgoner 
I'm okay with having a fp32 implementation for now.

Comment (xcgoner) [reply src/gluonnlp/optimizer/lamb.py]: @eric-haibin-lin I think currently only SGD has fp16 implementation, right? Adam and other optimizers have only fp32 implementation. So I think it's ok for now.

**Comment by mli** [comment]
Job PR-733/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-733/10/index.html

**Comment by mli** [comment]
Job PR-733/11 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-733/11/index.html
